### PR TITLE
Upload zipped pack to marketplce

### PR DESCRIPTION
**Pull request by ilappe**
## Status
- [x] In Progress
- [ ] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
https://github.com/demisto/etc/issues/33527
https://github.com/demisto/etc/issues/33528

## Description
New command to upload zipped pack to marketplace

## Must have
- [x] Tests
- [x] Documentation

### CODE REVIEW

**Comment by ilappe** [comment]
The build fail due to missing the right demisto-py
work on this

File: demisto_sdk/commands/upload/uploader.py
Comment (teizenman) [inline demisto_sdk/commands/upload/uploader.py:266]:
```suggestion
```

File: demisto_sdk/commands/upload/uploader.py
Comment (teizenman) [inline demisto_sdk/commands/upload/uploader.py:280]: Let's change the `0` and `1` to constants with indicative names, and change it all over this page.

**Review by teizenman — CHANGES_REQUESTED** [review]
1. We need to run a nightly build to test the changes in the `update_server_configuration` function.
2. did you test to see if deleting the configuration key makes the signature verification to be turned back on? Were you able to upload a non-signed pack after the deletion or not?

**Comment by ilappe** [comment]
> 1. We need to run a nightly build to test the changes in the `update_server_configuration` function.
> 2. did you test to see if deleting the configuration key makes the signature verification to be turned back on? Were you able to upload a non-signed pack after the deletion or not?

For sure, test & check in server side code
server search for such key to determine if verification is needed 
and if key not exist - he verify by default

File: demisto_sdk/__main__.py
Comment (ilappe) [reply demisto_sdk/__main__.py:276]: I used the `pop` because this popped arguments shouldn't pass to the `PackUnifier` 
and `pop` remove the from the kwargs

File: demisto_sdk/commands/compile_packs/packs_compiler.py
Comment (ilappe) [reply demisto_sdk/commands/compile_packs/packs_compiler.py:76]: This part was copied from the create_content_artifacts
and the motivation is to delete the files created during the process if the zip not finished successfully

File: demisto_sdk/commands/compile_packs/packs_compiler.py
Comment (ilappe) [reply demisto_sdk/commands/compile_packs/packs_compiler.py:79]: It's a code in create_content_artifact and i didn't want to touch
it's zip all the dirs into one zip file depended on flag 
i my case the flag is on and all will be as expected in one zip

File: demisto_sdk/commands/compile_packs/packs_compiler.py
Comment (ilappe) [reply demisto_sdk/commands/compile_packs/packs_compiler.py:83]: will delete the directories created during the process and not needed after zip was created

File: demisto_sdk/commands/common/content/objects/pack_objects/pack.py
Comment (ilappe) [reply demisto_sdk/commands/common/content/objects/pack_objects/pack.py:265]: Errors are well handled in the outer scope Uploader.zipped_pack_uploader
here the `try` `finally` is just to handle the cases we need to turn back the pack verification in server

File: demisto_sdk/commands/common/content/objects/pack_objects/pack.py
Comment (ilappe) [reply demisto_sdk/commands/common/content/objects/pack_objects/pack.py:33]: i delete the verification key flag in server when try to turn on the verification (delete the key == turn on)
and the reason why to delete it is because i don't want to leave not required key (key = true) as in most of cases the key was not exist before the command run
so in case of error the message ask the user to delete it manually

File: demisto_sdk/commands/compile_packs/packs_compiler.py
Comment (teizenman) [reply demisto_sdk/commands/compile_packs/packs_compiler.py:76]: But is it used in your code in any place?

File: demisto_sdk/commands/compile_packs/packs_compiler.py
Comment (teizenman) [reply demisto_sdk/commands/compile_packs/packs_compiler.py:79]: Did you try to install it on your machine? 
That won't work since it's expecting zips and not directories.

File: demisto_sdk/commands/unify_packs/packs_unifier.py
Comment (teizenman) [inline demisto_sdk/commands/unify_packs/packs_unifier.py:65]:
```suggestion
        """init dict that map pack name to Pack object and init also the pack_names property

        Args:
            pack_paths: CSV str with the pack paths

        """
```

This is the right way to write the docstring, please change the rest as well.
Look at the example here: https://gist.github.com/redlotus/3bc387c2591e3e908c9b63b97b11d24e#file-docstrings-py-L79

File: demisto_sdk/commands/unify_packs/packs_unifier.py
Comment (teizenman) [inline demisto_sdk/commands/unify_packs/packs_unifier.py:48]: Add docstring for the purpose of this class.
See example here: https://gist.github.com/redlotus/3bc387c2591e3e908c9b63b97b11d24e#file-docstrings-py-L154

File: demisto_sdk/commands/unify_packs/packs_unifier.py
Comment (teizenman) [inline demisto_sdk/commands/unify_packs/packs_unifier.py:22]: Why did you choose `cpus=os.cpu_count()`?

File: demisto_sdk/commands/unify_packs/packs_unifier.py
Comment (teizenman) [inline demisto_sdk/commands/unify_packs/packs_unifier.py:136]: @roysagi 
Please give your opinion on it.
If you think that's a good solution then maybe we should move it to `tools.py`?

File: demisto_sdk/commands/test_content/tools.py
Comment (teizenman) [inline demisto_sdk/commands/test_content/tools.py:19]:
```suggestion
        logging_manager (logging.Logger): Logging manager object
```

File: demisto_sdk/__main__.py
Comment (teizenman) [inline demisto_sdk/__main__.py]: I think that you should use the `PathsParamType` here: https://github.com/demisto/demisto-sdk/blob/master/demisto_sdk/__main__.py#L71

File: demisto_sdk/commands/common/content/objects/pack_objects/pack.py
Comment (teizenman) [inline demisto_sdk/commands/common/content/objects/pack_objects/pack.py:262]:
```suggestion
        logger.info('Turning off the server verification for signed packs')
```

Please change all the logger outputs to use capital letters at the beginning of the sentence.

File: demisto_sdk/commands/unify_packs/packs_unifier.py
Comment (teizenman) [inline demisto_sdk/commands/unify_packs/packs_unifier.py:95]: Just to make sure, you need this method in order to override the `super` class method, right?

File: demisto_sdk/commands/unify_packs/packs_unifier.py
Comment (teizenman) [inline demisto_sdk/commands/unify_packs/packs_unifier.py:130]: Why not putting it under the `with QuiteModeController`?

File: demisto_sdk/commands/unify_packs/tests/packs_unifier_test.py
Comment (teizenman) [inline demisto_sdk/commands/unify_packs/tests/packs_unifier_test.py:19]: Did you check that none of those constants exists in the `constants.py`/`test_constants.py`?

**Review by teizenman — CHANGES_REQUESTED** [review]
Please add type annotations for functions' arguments and return values.

File: demisto_sdk/commands/unify_packs/packs_unifier.py
Comment (ilappe) [reply demisto_sdk/commands/unify_packs/packs_unifier.py:130]: there is two kind of outputs
1. a lot of info printed by the `dump_pack` - will block in any case
2. a summery of **unified** resources  - will display in case of quite_mode = false

File: demisto_sdk/commands/unify_packs/packs_unifier.py
Comment (ilappe) [reply demisto_sdk/commands/unify_packs/packs_unifier.py:95]: Exactly

File: demisto_sdk/commands/unify_packs/packs_unifier.py
Comment (ilappe) [reply demisto_sdk/commands/unify_packs/packs_unifier.py:22]: is a mandatory argument in ArtifactsManager - i changed it to pass cpus=1 as we don't use multi processing

File: demisto_sdk/__main__.py
Comment (ilappe) [reply demisto_sdk/__main__.py]: In that implementation - the entire command will fail even if i have mistake in one pack path
i implement check that just ignore the invalid path and do the work on the valid packs
should i change it?

**Review by teizenman — CHANGES_REQUESTED** [review]

**Review by roysagi — CHANGES_REQUESTED** [review]
@ilappe - This looks very good!
Please see my comments.

Also we need to add this to the nightly work with server flow - for example https://app.circleci.com/pipelines/github/demisto/content/99307/workflows/5270e257-fad7-4628-a1d0-97a89b6570cd/jobs/395060#queue-placeholder/containers/0

File: demisto_sdk/commands/zip_packs/packs_zipper.py
Comment (ilappe) [reply demisto_sdk/commands/zip_packs/packs_zipper.py:30]: Why ?
it's not the same method
and even so - it's play like a manager that call to other methods and do something else in addition

File: demisto_sdk/__main__.py
Comment (teizenman) [reply demisto_sdk/__main__.py]: @roysagi Your input here is needed - what is the way we want to go here?

File: demisto_sdk/commands/upload/uploader.py
Comment (teizenman) [inline demisto_sdk/commands/upload/uploader.py:309]:
```suggestion
                            'Are you sure you want to continue? Y/[N]', fg='bright_red')
```
That's for indicating what is the default value.

File: demisto_sdk/commands/zip_packs/packs_zipper.py
Comment (roysagi) [reply demisto_sdk/commands/zip_packs/packs_zipper.py:30]: It is not instantly clear which `zip_packs` is being called in the `PacksDirsHandler` (which is in the same file) IMO.
I think that for future debugging purposes it might be wise to distinct this method from the imported one.

But it is not major - I leave it to @teizenman to decide the course of action here.

Comment (ilappe) [reply demisto_sdk/commands/zip_packs/packs_zipper.py:30]: Hi @roysagi 
@yaakovi say it's not real issue and can remain with this name
so please approve if you don't have any other comment
Thanks!!

**Review by roysagi — APPROVED** [review]

**Comment by ilappe** [comment]
Waiting for nightly build flow to be fixed in order validate this PR

**Comment by ilappe** [comment]
@teizenman please approve if no other comments 
thanks!

**Review by teizenman — APPROVED** [review]

**Review by teizenman — CHANGES_REQUESTED** [review]
Please verify that you completed all the comments that was given.

File: demisto_sdk/commands/compile_packs/packs_compiler.py
Comment (ilappe) [reply demisto_sdk/commands/compile_packs/packs_compiler.py]: Yes 
un case of error in middle the creation - the created dirs will be deleted

File: demisto_sdk/__main__.py
Comment (ilappe) [reply demisto_sdk/__main__.py]: Done

File: demisto_sdk/commands/create_artifacts/content_artifacts_creator.py
Comment (ilappe) [reply demisto_sdk/commands/create_artifacts/content_artifacts_creator.py:955]: Done 
also moved to `packs_zipper` without using the process pool

File: demisto_sdk/commands/upload/README.md
Comment (ilappe) [reply demisto_sdk/commands/upload/README.md]: done
waiting for review in https://github.com/demisto/content-docs/pull/722

File: demisto_sdk/commands/compile_packs/packs_compiler.py
Comment (teizenman) [reply demisto_sdk/commands/compile_packs/packs_compiler.py]: I meant to ask about `artifact_manager.exit_code = EX_FAIL`, is this exit code relevant in any place?

File: demisto_sdk/commands/compile_packs/packs_compiler.py
Comment (teizenman) [reply demisto_sdk/commands/compile_packs/packs_compiler.py]: ?

File: demisto_sdk/__main__.py
Comment (teizenman) [reply demisto_sdk/__main__.py]: ?

**Review by teizenman — APPROVED** [review]

**Pull request merged**
